### PR TITLE
add Ruby YAML unsafe_* methods as sink

### PR DESCRIPTION
**Pull request by amammad**

### CODE REVIEW

File: ruby/ql/src/experimental/CWE-502/YAMLUnsafeYamlDeserialization.ql [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    // for detecting The CVE we should uncomment following line instead of current RemoteFlowSource
+    // source instanceof DataFlow::LocalSourceNode
+    source instanceof UnsafeDeserialization::Source
+  }
+
+  override predicate isSink(DataFlow::Node sink) {
+    // for detecting The CVE we should uncomment following line
+    // sink.getLocation().getFile().toString().matches("%yaml_column%") and
+    sink instanceof YamlSink or
+    sink =
```
Comment (pwntester) [inline ruby/ql/src/experimental/CWE-502/YAMLUnsafeYamlDeserialization.ql:57]: Did you mean to place the sink at the receiver of the call?

File: ruby/ql/src/experimental/CWE-502/YAMLUnsafeYamlDeserialization.ql [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      API::getTopLevelMember(["YAML", "Psych"])
+          .getAMethodCall(["parse", "parse_stream", "parse_file"])
+          .getAMethodCall("to_ruby")
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
```
Comment (pwntester) [inline ruby/ql/src/experimental/CWE-502/YAMLUnsafeYamlDeserialization.ql:63]: Would it make sense to add these taint steps as global taint steps?

File: ruby/ql/src/experimental/CWE-502/YAMLUnsafeYamlDeserialization.ql (hunk 1 above)
Comment (pwntester) [inline ruby/ql/src/experimental/CWE-502/YAMLUnsafeYamlDeserialization.ql:57]: Make its own `YamlSink` for this. Would it make sense to add the sink at the reciever of the `to_ruby` call instead?

**Comment by aibaars** [comment]
@pwntester While the bug bounty policy says that new queries should go into the "experimental" folder, I think in this case the additional sinks could be added to the existing unsafe deserialization query. We could validate the additional sinks by running the existing and the updated query against a batch of say 1000 repos and see whether the new results are good. What do you think?

**Comment by pwntester** [comment]
Hi @amammad, as per @aibaars and @hmac suggestion, can you please add the new YAML/PLIST sinks into the existing `rb/unsafe-deserialization` query?

> Hi, I want to know about the load method that is safe now, but still some repositories are using the lower versions of Psych that load is unsafe. should codeql contains both of them as sink?

In most of the cases we remove a sink when it is no longer vulnerable in some version since Repos that use the old version should listen to dependabot an upgrade (or turn on dependabot in the first place).

File: ruby/ql/src/experimental/CWE-502/YAMLUnsafeYamlDeserialization.ql (hunk 2 above)
Comment (pwntester) [reply ruby/ql/src/experimental/CWE-502/YAMLUnsafeYamlDeserialization.ql]: The taintsteps defined in the `isAdditionalTaintStep` of a TaintTracking configuration only apply to that particular configuration. I can imagine untrusted YAML data being parsed and then one of the properties flowing into a different injection sink (eg: SQL injection). In those cases we want to apply the taintsteps to all Tracking configurations. To do that, create a class that extends from `AdditionalTaintStep`:

```ql
private import codeql.ruby.dataflow.FlowSteps

private class YamlParseStep extends AdditionalTaintStep {
  override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
    ...
  }
}
```

You can see some examples [here](https://github.com/github/codeql/blob/main/ruby/ql/lib/codeql/ruby/frameworks/StringFormatters.qll). You would need to create a Yaml.qll file in `ruby/ql/lib/codeql/ruby/frameworks` and then import it from https://github.com/github/codeql/blob/main/ruby/ql/lib/codeql/ruby/Frameworks.qll

@hmac @aibaars can you verify we want to do that?

**Comment by pwntester** [comment]
> I have an important question, I want to submit bug slayer request for this query too, if i merged it in a non-experimental directory so all GitHub repositories that are using codeql will recognize the vulnerability and I can't really submit my bug slayer request as maintainers will fix the vulnerabilities according to codeql action alerts.

It will take some time to merge this PR and make it available for Code Scanning. In the meantime you can mass scan for this issue using MRVA and report any findings to the maintainers.

File: ruby/ql/lib/codeql/ruby/security/UnsafeDeserializationQuery.qll [hunk 3]
```diff
@@ -23,6 +24,27 @@ class Configuration extends TaintTracking::Configuration {
 
   override predicate isSink(DataFlow::Node sink) { sink instanceof UnsafeDeserialization::Sink }
 
+  override predicate isAdditionalTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
```
Comment (pwntester) [inline ruby/ql/lib/codeql/ruby/security/UnsafeDeserializationQuery.qll:27]: Please define these as global taint steps

**Comment by calumgrant** [comment]
Closing in favour of https://github.com/github/codeql/pull/13307

**Pull request closed without merging**